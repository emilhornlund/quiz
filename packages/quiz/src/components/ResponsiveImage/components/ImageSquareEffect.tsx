import { CountdownEvent } from '@quiz/common'
import React, { useEffect, useRef, useState } from 'react'

interface ImageEffectProps {
  box: { w: number; h: number }
  countdown?: CountdownEvent
  numberOfSquares: number
}

export const ImageSquareEffect: React.FC<ImageEffectProps> = ({
  box,
  countdown,
  numberOfSquares,
}) => {
  const evenSquares =
    numberOfSquares % 2 === 0 ? numberOfSquares : numberOfSquares + 1

  const [visibleSquares, setVisibleSquares] = useState(evenSquares - 1)
  const intervalRef = useRef<NodeJS.Timeout | null>(null)

  useEffect(() => {
    if (intervalRef.current) {
      clearInterval(intervalRef.current)
      intervalRef.current = null
    }

    if (!countdown) return

    const serverTime = new Date(countdown.serverTime).getTime()
    const initiated = new Date(countdown.initiatedTime).getTime()
    const expiry = new Date(countdown.expiryTime).getTime()

    if ([serverTime, initiated, expiry].some((t) => isNaN(t))) {
      setVisibleSquares(0)
      return
    }

    const clientToServerOffset = serverTime - Date.now()
    const totalDuration = expiry - initiated
    if (totalDuration <= 0) {
      setVisibleSquares(0)
      return
    }

    const updateSquares = () => {
      const now = Date.now() + clientToServerOffset
      const elapsed = now - initiated
      const progress = Math.min(elapsed / totalDuration, 1)

      const removed = Math.floor(progress * evenSquares) + 1
      setVisibleSquares(Math.max(evenSquares - removed, 0))
    }

    updateSquares()
    intervalRef.current = setInterval(updateSquares, 200)

    return () => {
      if (intervalRef.current) {
        clearInterval(intervalRef.current)
        intervalRef.current = null
      }
    }
  }, [countdown, evenSquares])

  return (
    <div
      style={{
        position: 'absolute',
        top: 0,
        left: 0,
        width: box.w,
        height: box.h,
        display: 'flex',
        flexWrap: 'wrap',
        alignContent: 'flex-start',
      }}>
      {Array.from({ length: evenSquares }).map((_, index) => (
        <div
          key={index}
          style={{
            width: `${box.w / (evenSquares / 2)}px`,
            height: `${box.h / 2}px`,
            background: index < visibleSquares ? 'white' : 'transparent',
          }}
        />
      ))}
    </div>
  )
}
