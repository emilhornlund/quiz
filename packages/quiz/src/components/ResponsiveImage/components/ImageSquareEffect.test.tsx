import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import { ImageSquareEffect } from './ImageSquareEffect'

const defaultProps = {
  box: { w: 200, h: 100 },
  numberOfSquares: 4,
  countdown: undefined,
}

describe('ImageSquareEffect', () => {
  it('renders correctly (snapshot)', () => {
    const { container } = render(<ImageSquareEffect {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  it('renders with countdown (snapshot)', () => {
    const countdown = {
      serverTime: '2025-10-12T12:00:00.000Z',
      initiatedTime: '2025-10-12T11:59:59.000Z',
      expiryTime: '2025-10-12T12:00:01.000Z',
    }
    const { container } = render(
      <ImageSquareEffect {...defaultProps} countdown={countdown} />,
    )
    expect(container).toMatchSnapshot()
  })
})
