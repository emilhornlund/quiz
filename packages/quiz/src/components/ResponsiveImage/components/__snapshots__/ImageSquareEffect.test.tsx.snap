// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageSquareEffect > renders correctly (snapshot) 1`] = `
<div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 200px; height: 100px; display: flex; flex-wrap: wrap; align-content: flex-start;"
  >
    <div
      style="width: 100px; height: 50px; background: white;"
    />
    <div
      style="width: 100px; height: 50px; background: white;"
    />
    <div
      style="width: 100px; height: 50px; background: white;"
    />
    <div
      style="width: 100px; height: 50px; background: transparent;"
    />
  </div>
</div>
`;

exports[`ImageSquareEffect > renders with countdown (snapshot) 1`] = `
<div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 200px; height: 100px; display: flex; flex-wrap: wrap; align-content: flex-start;"
  >
    <div
      style="width: 100px; height: 50px; background: white;"
    />
    <div
      style="width: 100px; height: 50px; background: transparent;"
    />
    <div
      style="width: 100px; height: 50px; background: transparent;"
    />
    <div
      style="width: 100px; height: 50px; background: transparent;"
    />
  </div>
</div>
`;
